Clean/Table/Basic.lean

```diff
@@ -200,110 +135,118 @@ namespace CellOffset
   Current row offset
 -/
 @[table_norm]
-def curr {W : ℕ+} {S : Type -> Type} [struct: ProvableType S] (j : Fin (struct.size)) :  CellOffset W S := ⟨0, j⟩
+def curr {W : ℕ+} (j : Fin (size S)) :  CellOffset W S := ⟨0, j⟩
 
 /--
   Next row offset
 -/
 @[table_norm]
-def next {W : ℕ+} {S : Type -> Type} [struct: ProvableType S] (j : Fin (struct.size)) :  CellOffset W S := ⟨1, j⟩
+def next {W : ℕ+} (j : Fin (size S)) :  CellOffset W S := ⟨1, j⟩
 
 end CellOffset
 
 /--
   Mapping from the index of a variable to a cell offset in the table.
 -/
 @[reducible]
-def CellAssignment (W: ℕ+) (S : Type -> Type) [ProvableType S] := ℕ -> CellOffset W S
+def CellAssignment (W: ℕ+) (S : Type → Type) [ProvableType S] := ℕ → CellOffset W S
 
 /--
   Atomic operations for constructing a table constraint, which is a constraint applied to a window
   of rows in a table.
 -/
-inductive TableConstraintOperation (W : ℕ+) (S : Type -> Type) (F : Type) [Field F] [struct: ProvableType S] where
+inductive TableConstraintOperation (W : ℕ+) (S : Type → Type) (F : Type) [Field F] [ProvableType S] where
   /--
     Add some witnessed variable to the context
   -/
-  | Witness : CellOffset W S -> (compute : Unit → F) -> TableConstraintOperation W S F
+  | Witness : CellOffset W S → (compute : Unit → F) → TableConstraintOperation W S F
 
   /--
     Witness a fresh variable for each cell in the row at some offset `off` in the trace
   -/
-  | GetRow : (off : Fin W) -> TableConstraintOperation W S F
+  | GetRow : (off : Fin W) → TableConstraintOperation W S F
 
   /--
     Allocate a subcircuit in the trace
   -/
-  | Allocate: {n: ℕ} → SubCircuit F n -> TableConstraintOperation W S F
+  | Allocate: {n: ℕ} → SubCircuit F n → TableConstraintOperation W S F
 
   /--
     Assign a variable to a cell in the trace
   -/
-  | Assign : Variable F -> CellOffset W S -> TableConstraintOperation W S F
+  | Assign : Variable F → CellOffset W S → TableConstraintOperation W S F
 
 /--
   Context of the TableConstraint that keeps track of the current state, this includes the underlying
   offset, and the current assignment of the variables to the cells in the trace.
 -/
-structure TableContext (W: ℕ+) (S : Type -> Type)  (F : Type) [Field F] [struct: ProvableType S] where
+structure TableContext (W: ℕ+) (S : Type → Type) (F : Type) [Field F] [ProvableType S] where
   offset: ℕ
   assignment : CellAssignment W S
+  operations: List (TableConstraintOperation W S F)
```

Comment:
table context is the monad state, so it must now also carry along the `operations` done so far